### ISSUE TITLE
Standard schema for results, including errors

### ISSUE THREAD

**Issue by trajano** [json-schema-org/json-schema-spec#396]
The schema provides a way of validating the input, but there should be a standard for the validation results as well.

**Comment by handrews** [comment]
@trajano could you elaborate on this a bit more?  Are you talking about input to a specific validator?  Or input/results from some other system that you want to use validation with?

**Comment by trajano** [comment]
I was thinking that we should have something along the lines of a validation result structure akin to https://docs.oracle.com/javaee/7/api/javax/validation/ConstraintViolation.html

For example:
```
{
   "valid": true | false,
   "violations": [
      {
         "property-path" : "$/path/to[10]/bad/element",
         "invalid-value": "who's bad" | { "some": [ "complex", "object" ] }
         "rule-path": "$/path/in/schema/that/triggered/violation",
          // these may be optionally added but should not be part of the spec
         "message-template": "{0} is who is bad",
         "message": "who's bad is who is bad",
     } ,
      {
         "property-path" : "$/path/to[10]/bad/element",
         "invalid-value": "who's bad" | { "some": [ "complex", "object" ] }
         "rule-path": "$/path/in/schema/that/triggered/violation",
          // these may be optionally added but should not be part of the spec
         "message-template": "{0} is who is bad",
         "message": "who's bad is who is bad",
     } 
    ]
}
```

**Comment by trajano** [comment]
That way each application need not develop their own validation error standard messages

**Comment by handrews** [comment]
@trajano I like the idea, in some form or another.

RFCs generally do not constrain implementations in this way, due to the wide variety of implementation languages, environments, and requirements such as performance vs usability.  This idea helps interoperability among users of validators, rather than interoperability in schema processing and outcome.  So I don't think this proposal belongs in the spec repo (at least not with its current scope).

I see two possible approaches:

1.  Write a schema for the desired output and perhaps we can discuss it with implementation maintainers and publish it as a recommendation on the web site (which is a separate GitHub repo).  Depending on how broadly it is adopted maybe there will be a clear need to formalize it further.

2.  Approach it like the `"application/problem+json"` specification.  That is an error-reporting format that is defined independent of the system producing the errors.  You would obviously want specific schema error-oriented fields, and would not be going for the completely generic approach of that spec, but separating it encourages its use in implementations that can support it, without burdening more constrained implementations with support.

**Comment by erayd** [comment]
I really like this:
> 2. Approach it like the "application/problem+json" specification. That is an error-reporting format that is defined independent of the system producing the errors. You would obviously want specific schema error-oriented fields, and would not be going for the completely generic approach of that spec, but separating it encourages its use in implementations that can support it, without burdening more constrained implementations with support.

If it's attractive enough, I would certainly support it in the implementations I'm involved with.

**Comment by trajano** [comment]
I think `application/violation+json` as a MIME type sounds better.  For those who may not know (like me) the thing before the `+json` is the MIME subtype.

**Comment by handrews** [comment]
@trajano @erayd I'm glad to see multiple people taking interest.  I do not have time to focus on this right now, but here is what I would recommend:

* Don't worry about making it an actual media type right now.  You may want to do that, or you may (for instance) just want to define a specific extension of `application/problem+json` that can be identified by a schema describing the validation error output.  Then you do not need to go through RFC, but can instead just apply two existing technologies and see how that goes to start.
* File an issue to track this over in the json-schema-org.github.io repo.  Storing your proposal there as extended documentation / best practices seems like the best starting point to me.
* I'm going to close this as I do not think it belongs in this repo at this time.  But that is just me managing the repo scope, not a rejection of the idea at all.

**Comment by trajano** [comment]
Closing this issue since it is out of scope of this project.

**Comment by yurikhan** [comment]
I am interested in this. Subscribing for comments, even if only to know which other issue to track.

**Comment by handrews** [comment]
There have been a few other issues around error reporting and/or formal definitions of validation results, so I have made an "output" label to track these.  I will re-open this and label it, at least until we figure out where all of these concerns should really live.

Somewhat related: I'm defining a "recommended" output format for Hyper-Schema, although it is not a strict requirement.  It does seem useful, though, which is why I am re-considering whether a recommendation, if not requirement, might be in-scope.  The situation in hyper-schema is a little different, though (the mandatory output of the process is much more complex than a boolean result, even ignoring error reporting).

**Comment by Anthropic** [comment]
@handrews should validation results be a vocabulary? I think it would be awesome to have a consistent format/structure/schema for responses that validation implementations can work toward to ensure compatibility to make transferring from one validator to another more seamless. Just a week ago I posted [this issue](https://github.com/korzio/djv/issues/54) on [djv](https://github.com/korzio/djv) to support a similar, more detailed, [error message format](https://github.com/epoberezkin/ajv#validation-errors) to [ajv](https://github.com/epoberezkin/ajv). I think if [ajv](https://github.com/epoberezkin/ajv) and [djv](https://github.com/korzio/djv) the two fastest validators have a similar format then it could act as a starting point for a predictable validation result schema.

**Comment by handrews** [comment]
@Anthropic I'm not sure I would consider output formats to be "vocabularies", as to me a vocabulary is a set of keywords to use in schemas to either annotate or assert conditions about an instance document.  I think output documents are instances (and not also schemas).  They can be described by a schema, but they do not constitute a schema as far as I can tell.

Unless I am misunderstanding what you mean?  I don't think we need a new vocabulary to write a schema describing the output format.

When you look at the hyper-schema rewrite, you will see the definition of the output format there.  Although it does not include an error reporting format.

I have no particular point here, just acknowledging that there is something to this topic and I'm not sure where it fits.

**Comment by Anthropic** [comment]
@handrews I see your non particular point :wink:, what I meant was it just felt looking at the ajv error format with `keyword`, `dataPath`, `schemaPath`, `params` and `message` seemed to have its own keywords, so I wasn't sure where it belongs, you could say I was thinking out loud. Not concerned with the where as much as wanting to figure out or get ideas on the best place to define such a format.

**Comment by gregsdennis** [comment]
I'm already doing something similar in my implementation, though it's merely in an object model.  I'd have to update that model and make it serializable, but it's not a stretch for me to support something like this.

I like the idea of standardizing output.

**Comment by vearutop** [comment]
Important thing is that validation failures can be deeply nested by `*Of` keywords. And to understand the root error you'll need to follow sub-errors.
Sample [error message](https://github.com/swaggest/php-json-schema#understanding-error-cause):
```
No valid results for oneOf {
 0: Enum failed, enum: ["a"], data: "f" at #->properties:root->patternProperties[^[a-zA-Z0-9_]+$]:zoo->oneOf[0]
 1: Enum failed, enum: ["b"], data: "f" at #->properties:root->patternProperties[^[a-zA-Z0-9_]+$]:zoo->oneOf[1]
 2: No valid results for anyOf {
   0: Enum failed, enum: ["c"], data: "f" at #->properties:root->patternProperties[^[a-zA-Z0-9_]+$]:zoo->oneOf[2]->$ref[#/cde]->anyOf[0]
   1: Enum failed, enum: ["d"], data: "f" at #->properties:root->patternProperties[^[a-zA-Z0-9_]+$]:zoo->oneOf[2]->$ref[#/cde]->anyOf[1]
   2: Enum failed, enum: ["e"], data: "f" at #->properties:root->patternProperties[^[a-zA-Z0-9_]+$]:zoo->oneOf[2]->$ref[#/cde]->anyOf[2]
 } at #->properties:root->patternProperties[^[a-zA-Z0-9_]+$]:zoo->oneOf[2]->$ref[#/cde]
} at #->properties:root->patternProperties[^[a-zA-Z0-9_]+$]:zoo
```

I'm not sure standard way of error reporting is necessary (the only obvious use-case to me is to change one implementation to another and keep same error handling code), but I would implement it.

**Comment by handrews** [comment]
@vearutop yeah, the `*Of` keywords are a major concern.  I think the key for them is designing a simple raw error format that tools can navigate to provide better feedback.  For example, being able to show errors in some sort of hierarchical drill-down display.

This is off the top of my head and not a serious proposal:

If I'm reading your example right you're working with a schema like the following:

```JSON
{
  "type": "object",
  "properties": {
    "root": {
      "type": "object",
      "patternProperties": {
        "^[a-zA-Z0-9_]+$": {
          "oneOf": [
            {"enum": ["a"]},
            {"enum": ["b"]},
            {"$ref": "#/cde"}
          ]
        }
      }
    }
  },
  "cde": {
    "anyOf": [
      {"enum": ["c"]},
      {"enum": ["d"]},
      {"enum": ["e"]}
    ]
  }
}
```

with an instance of:

```JSON
{
  "root": {
    "zoo": "f"
  }
}
```

An error data structure could look something like:

```JSON
[
  {
    "instanceLocation": "/root/zoo",
    "instanceData": "f",
    "errors": [
      {
        "schemaLocation": ["#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf"],
        "validSubschemas": [],
        "subschemaErrors": [
          {
            "schemaLocation": ["#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf/0/enum"],
            "schemaValue": ["a"]
          },
          {
            "schemaLocation": ["#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf/1/enum"],
            "schemaValue": ["b"]
          },
          {
            "schemaLocation": ["#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf/2/$ref", "/cde/anyOf"],
            "validSubschemas": [],
            "subschemaErrors": [
              {
                "schemaLocation": ["#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf/2/$ref", "#/cde/anyOf/0/enum"],
                "schemaValue": ["c"]
              },
              {
                "schemaLocation": ["#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf/2/$ref", "#/cde/anyOf/1/enum"],
                "schemaValue": ["d"]
              },
              {
                "schemaLocation": ["#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf/2/$ref", "#/cde/anyOf/2/enum"],
                "schemaValue": ["e"]
              }
            ]
          }
        ]
      }
    ]
  }
]
```

This organizes errors first by the location in the instance that is failing validation, then indicates each schema location against which it fails that may be the cause of the overall failure.

Instance locations are plain (non-URI-fragment) JSON Pointers, while schema locations are arrays of URIs with JSON pointer fragments where a new pointer is added to the array whenever you cross a `$ref`.

If any of the subschemas were valid, `validSubschemas` would provide the URIs with JSON Pointer fragments of those subschemas.  In this case, none are, so it's an empty array.

The goal here is to provide all of the key data in a standardized way, but not necessarily the natural language phrasing.  So there's nothing here that says "Value not present in enum", because there are many ways that that phrasing might be chosen, not to mention I18N/L10N/A11Y issues.

But this approach would let tools display something simple like "No valid `oneOf` subschemas at..." and only display each individual error when drilling down.  In my experience, the biggest problem with `*Of` error reporting is that it's hard to sort through everything to figure out what the real problem is vs "errors" that are expected (particularly with `oneOf` where all but one *should* fail).

**Comment by handrews** [comment]
Since draft-08 is intended to include annotation collection, which would involve a recommended output format, we should really handle a recommended error format at the same time.  That would cover all of this issue, I believe, so I'm adding it to the milestone.

**Comment by vearutop** [comment]
I think it is worth adding `keyword` to error data structure. 
@handrews I've tried to describe your example (with few changes) in JSON Schema: https://gist.github.com/vearutop/17ef696fe1426844b302e844076400c5

**Comment by wichert** [comment]
@handrews I like your approach. From a UI perspective it looks like it provides all information necessary to produce useful error message, without leaking any implementation details.

There is a very nice quality to your proposal: it gives a very convenient way to define good error messages putting error messages directly in schema and looking them up using `schemaLocation` so you can do something like this:

```json
{
  "type": "object",
  "properties": {
    "email": {
      "type": "string",
      "format": "email",
      "error": {
        "required": "Please enter your personal email address",
        "format": "Please a valid email address."
      }
   }
}
```

If `schemaLocation` is `#/properties/email/format` the error message is `#/propertiers/email/errors/format`. You can of course keep the error messages in a separate structure as well instead of polluting the schema. This is well out of scope for the current discussion, but I can see it being very attractive.

**Comment by wichert** [comment]
I realise that in my previous comment I made an assumption that you can just remove the last part of the JSON Pointer in `schemaLocation` to get a pointer to the variable. Is that a valid assumption, or would be it useful to add a `variableLocation` key?

**Comment by handrews** [comment]
@wichert 

> it provides all information necessary to produce useful error message, without leaking any implementation details.

Thats a great way to summarize one of the key use cases.  We don't want to tell people how to _display_ errors, and we don't want display code to have to understand how any one specific validator (or hyper-schema client, or whatever) works.  It may help to understand how they work in a general sense, but it should be easy for someone to write an error display library independent of any validator.

Reporting errors seems to be one of the hardest parts of writing a validator, and this could separate those concerns.

> I realise that in my previous comment I made an assumption that you can just remove the last part of the JSON Pointer in `schemaLocation` to get a pointer to the variable. Is that a valid assumption, or would be it useful to add a `variableLocation` key?

By "variable" do you mean the name of the object property with the value that was in error?  That should be apparent from `instanceLocation`, which is up at the top of my example.  There may be multiple errors for multiple schema locations for a single instance location.

-----

Regarding error strings or other messages, see also #148 (Add "messages" property) and #270 (Support for error "level").  I've long been skeptical of both as a part of the standard, but I suppose having a standard output/error format may make this sort of thing fit better.  I think they also work very well as extension properties (even if the implementation does not use them, when you go to look at the schema you'll see them).  And it's not clear where they go- core because they're a very general thing? 
 Validation because it's where we have assertions for the most part?  And how keyword-specific would message structures end up being?

So I'd probably prefer to leave messages and levels to extensions for now and possibly adopt them into a spec if a clear convention emerges.

**Comment by handrews** [comment]
@vearutop thanks for writing up the schema!  I actually had a `keyword` property in my example at first, and then had some concern over not all schema locations corresponding directly to a keyword.  

But thinking about it more (I wrote the example very quickly) I agree that most if not all are related to a keyword (e.g. ".../oneOf/2" is related to `oneOf` even if that schema location is only part of the `oneOf` value), and having to find the most relevant keyword in a JSON Pointer is annoying.

I've only had a chance to skim over the other details of your alterations but I like the direction.  We don't want to pin things down too much because the error format should work for extension keywords or future vocabularies without having to update it for the specific additional keywords involved.  But I think we can categorize keywords in a way that strikes a balance between listing them out and not being able to indicate different error formats (such as subschema errors) at all.

**Comment by Julian** [comment]
For reference, this is what the Python implementation uses here:

http://python-jsonschema.readthedocs.io/en/latest/errors/

Though yeah not in love with the names we ended up picking.

Will have to compare that to the proposed schema above though I'm pretty sure we at least cover the same information.

**Comment by yurikhan** [comment]
For another reference, here’s the description of what I implemented for RapidJSON: http://rapidjson.org/md_doc_schema.html#Reporting

**Comment by gregsdennis** [comment]
With some direction from @handrews, I think I have a format that addresses all of the primary concerns while also considering the advent of annotation collection.

In short, we want a format that

- Is easy to navigate
- Combines the instance data with the erroring/annotating keyword that pertains to it
- Provides paths to keywords that produce errors/annotations to be expressed in two formats
    - relative URI to indicate how the keyword was reached
    - absolute URI to provide a direct link (so that applications don't have to follow a dereference chain)

To address the first, the format mimics the structure of the JSON instance being validated.

To address the second and third, a recursive output format has been created that contains

- the instance data
- the overall validation result
- any errors in the event of a failure
- any annotations in the event of a success

## Examples

### Schema (for both examples)

```json
{
  "type":"object",
  "title":"root",
  "$defs":{
    "integerMin5":{
      "title":"referenced title",
      "type":"integer",
      "minimum":5
    }
  },
  "properties":{
    "intProp":{
      "title":"it's an int",
      "description":"found the int",
      "type":"integer",
      "minimum":5
    },
    "nested":{
      "type":"object",
      "title":"internal",
      "anyOf":[
        {
          "title":"a passed",
          "required":["a"]
        },
        {
          "title":"b passed",
          "required":["b"]
        }
      ]
    },
    "offsetCoordinate":{
      "type":"object",
      "title":"an offset coordinate",
      "properties":{
        "X":{"$ref":"/$defs/integerMin5"},
        "Y":{"$ref":"/$defs/integerMin5"}
      }
    }
  }
}
```

### Instances & Results

A valid instance

```json
{
  "intProp":9,
  "nested":{
    "b":[]
  }
}
```

produces the output

```json
{
  "instanceData":{
    "intProp":{
      "instanceData":9,
      "result":"passed",
      "annotations":[
        {
          "source":"/properties/intProp/title",
          "value":"it's an int"
        },
        {
          "source":"/properties/intProp/description",
          "value":"found the int"
        }
      ]
    },
    "nested":{
      "instanceData":{
        "b":{
          "instanceData":[],
          "result":"passed",
          "annotations":[
            {
              "source":"/properties/nested/anyOf/1/title",
              "value":"b passed"
            }
          ]
        }
      },
      "result":"passed",
      "annotations":[
        {
          "source":"/properties/nested/title",
          "value":"internal"
        }
      ]
    }
  },
  "result":"passed",
  "annotations":[
    {
      "source":"/title",
      "value":"root"
    }
  ]
}
```

A failing instance

```json
{
  "intProp":9,
  "offsetCoordinate":{
    "X":10,
    "Y":3
  },
  "otherProp":true
}
```

produces the output

```json
{
  "instanceData":{
    "intProp":{
      "instanceData":9,
      "result":"passed"
    },
    "offsetCoordinate":{
      "instanceData":{
        "X":{
          "instanceData":10,
          "result":"passed"
        },
        "Y":{
          "instanceData":3,
          "result":"failed",
          "errors":[
            {
              "source":"/properties/offsetCoordinate/Y/$ref/minimum",
              "absoluteSource":"/$defs/integerMin5",
              "message":"The value '3' is not greater than or equal to '10'."
            }
          ]
        }
      },
      "result":"failed",
      "errors":[
        {
          "source":"/properties/offsetCoordinate",
          "message":"A subschema failed validation."
        }
      ]
    },
    "otherProp":{
      "instanceData":true,
      "result":"noRequirements"
    }
  },
  "result":"failed",
  "errors":[
    {
      "source":"/",
      "message":"A subschema failed validation."
    }
  ]
}
```

***Note 1** All annotations are dropped when the overall schema fails, even those annotations for the portions of the instance that passed.*

***Note 2** The "A subschema failed validation" error may or may not be required.  It may be sufficient to merely include errors on those elements that explicitly failed.  This may be an implementation option.*

## Description of the output

The basic result object is recursive and can have the following properties:

- `instanceData` - The value of the instance at the current location.  Operates recursively to handle objects and arrays.
- `result` - An enumeration of `passed`, `failed`, or `noRequirements`
- `annotations` - *Only present when `result` is `passed`.* An array of annotation objects.  Each annotation object contains:
    - `source` - A JSON Pointer to the schema keyword that produced the annotation, relative from the root of the schema.
    - `absoluteSource` - If `source` contains a `$ref`, this is an absolute URI to the source of the annotation.
    - `value` - The annotation value.
- `errors` - *Only present when `result` is `failed`.* An array of error objects.  Each error object contains:
    - `source` - A JSON Pointer to the schema keyword that failed validation, relative from the root of the schema.
    - `absoluteSource` - If `source` contains a `$ref`, this is an absolute URI to the source of the failed keyword.
    - `message` - The error message.

Whenever a value in the JSON instance does not have any requirements, `result` will be `noRequirements`, and any sub-data of the instance is ignored (so there is no recursion into these values).  Additionally, neither `errors` nor `annotations` will be present in the result object.

For `instanceData`, when validation takes place or annotations collected, recursion occurs in the case of arrays and objects.  For arrays, each item in the array is encapsulated in the result object.  For objects, the keys are maintained and the value for each key is encapsulated in the result object.  This will result in similar navigation between the original JSON instance and its validation results object.  For example, if a path to a value in the original instance is

    /a/b/3/c

the corresponding validation results can be found in the results object at the path

    /instanceData/a/instanceData/b/instanceData/3/instanceData/c

*This is easily transformable by prepending/omitting `/instanceData` for each segment of the pointer path.*

`absoluteSource` only appears when `source` contains one or more `$ref`s, otherwise these values will be the same.  The absolute URI could point to a definition inside the same schema or to an external schema source.

## Output schema

As a bonus, here's a schema that describes the results object:

```json
{
  "type":"object",
  "$defs":{
    "base":{
      "type":"object",
      "properties":{
        "instanceData":{
          "oneOf":[
            {"type":["string","integer","number","boolean","null"]},
            {
              "type":"array",
              "items":{"$ref":"#"}
            },
            {
              "type":"object",
              "additionalProperties":{"$ref":"#"}
            }
          ]
        }
      },
      "required":["instanceData"]
    },
    "annotation":{
      "type":"object",
      "properties":{
        "source":{
          "type":"string",
          "format":"uri"
        },
        "absoluteSource":{
          "type":"string",
          "format":"uri"
        },
        "value":true
      },
      "required":["source","value"]
    },
    "passed":{
      "type":"object",
      "properties":{
        "result":{"const":"passed"},
        "annotations":{
          "type":"array",
          "items":{"$ref":"#/$defs/annotation"}
        }
      },
      "required":["result"]
    },
    "error":{
      "type":"object",
      "properties":{
        "source":{
          "type":"string",
          "format":"uri"
        },
        "absoluteSource":{
          "type":"string",
          "format":"uri"
        },
        "message":{"type":"string"}
      },
      "required":["source","message"]
    },
    "failed":{
      "type":"object",
      "properties":{
        "result":{"const":"failed"},
        "errors":{
          "type":"array",
          "items":{"$ref":"#/$defs/error"}
        }
      },
      "required":["result", "errors"]
    },
    "noRequirements":{
      "type":"object",
      "properties":{
        "result":{"const":"noRequirements"}
      }
    }
  },
  "allOf":[
    {"$ref":"#/$defs/base"}
  ],
  "oneOf":[
    {"$ref":"#/$defs/passed"},
    {"$ref":"#/$defs/failed"},
    {"$ref":"#/$defs/noRequirements"},
  ]
}
```

**Comment by trajano** [comment]
I both like and dislike the success returning processing results. 

I like it because I can see the processing results in case I have a false positive on the schema. 

I dislike it because it will use up space where most cases would be happy path.

Perhaps implementers need to make this configurable

**Comment by gregsdennis** [comment]
@Relequestual @epoberezkin @awwright You guys are notably quiet on this issue.  Have ye any thoughts?

**Comment by awwright** [comment]
I don't think it makes sense to define any sort of implementation details. We're defining a media type, not an API.

**Comment by gregsdennis** [comment]
@awwright we're trying to make the validation/annotation results part of the spec (or at least a parallel spec).

As far as my implementation goes, I would create an object model to represent the above structure (or whatever structure results from this discussion) that would then serialize appropriately.

The point is that output varies between implementations, and we're trying is standardize it.  This issue is less about a media type (yes it's mentioned above), and more about a standard output format for validation results and annotations.

**Comment by gregsdennis** [comment]
@handrews would you apply the **annotation** label to this as well, please?

**Comment by handrews** [comment]
@gregsdennis label added

@awwright There are two things going on here:

1.  Annotation collection, which has been a major focus for draft-08 for many, many reasons.  Hyper-Schema is essentially a complex annotation.  Code generation, UI generation, documentation generation- all done with annotations.  We need to provide guidance on what annotation output looks like or else applications will not be able to reliably make use of them.  We already provide a RECOMMENDED output format for hyper-schema.  It is my intention for this output to also be RECOMMENDED rather than required.

2.  One of the concerns raised when we spoke with the IETF JSON working group was that the implementations they had seen had inconsistent and usually unsatisfactory error reporting.  This was coming from one of the more helpful people, so I am taking that feedback seriously.  There have been many less formal complaints about the difficulties around error reporting, so providing a recommendation on what needs to be reported will help many thing.  Again, this would be RECOMMENDED rather than MUST.

It should be noted that an implementation is not required to collect annotations, and should not necessarily be required to produce errors (although unless there is a truly compelling reason why the target environment makes error reporting infeasible, I doubt such an implementation would get used much).

-----------

All: I am under the weather with a bad cold and will not be contributing much to this discussion for the next few days.  Other stuff that I've been posting or commenting on here is stuff that was already mostly complete (e.g. half-finished PRs that I just tidied up and pushed), or that I'd done the hard thinking about already and just needed a simple write-up.

In any event, I think that this is a topic that is best addressed by those who implement the spec, or who are implementing applications on top of spec implementations.  So I hope to see more commentary comparing proposals / existing output and error formats from @Julian @yurikhan @erayd @erosb and whoever else I'm forgetting.

Let's set aside the question of whether this output format is within the scope of the spec- it is obviously of great interest to many people, and the error side at least is known to be of interest to influential IETF people.  Whether this goes in the spec or as a less formal recommendation on the web site is irrelevant for now, let's just figure out what we want to recommend.

**Comment by handrews** [comment]
Also paging @mokkabonna @johandorland @korzio @davishmcclurg

**Comment by handrews** [comment]
@gregsdennis @yurikhan @Julian @trajano @vearutop you have all made or substantially commented on proposals, or shown what your implementation is doing in this area.

I don't have time right now to mediate a discussion, but as someone who does not write an implementation, I don't think I'm the most important person here anyway.  If you can agree, or come close to an agreement, on what this should look like, that would be of tremendous use to the project.  I'm also asking @philsturgeon if he can help moderate, although I'm not expecting him to add more proposals (unless he really wants to).

As noted earlier, we don't need to worry about whether this is part of the spec or a less formal recommendation on the web site, there is plenty of demand for this either way.  I'd just like for the community to settle on something while I work on all of the other things going into this next draft.

**Comment by Anthropic** [comment]
I started a [validator wrapper](https://github.com/jsonschema/assimilate) to ensure it was easier to switch validators in future if needed and @korzio was awesome enough to start changing [djv](https://github.com/korzio/djv/issues/54) errors in the direction of [ajv's current error format](https://github.com/epoberezkin/ajv#validation-errors) for consistency.

My feedback would be:
- A minimal output option for performance along with more verbose would be beneficial.
  
  Sometimes the end user wants to "stop on first error", it would be worth considering how that could or should look as well perhaps, if there is a recursive result.
- Recursing a tree doesn't give quick access to error counts where an array of **EDO**s (Error Description Objects) that conform to the requirements @gregsdennis laid out with instanceData etc.. would.
  
  A recursive tree output for a massive schema with hundreds of fields would be quite a performance burden to report back for an error on one field, how can that be addressed? Was it already and I missed it?

Thanks @handrews & @gregsdennis for continuing to draw attention to this issue! :)

**Comment by gregsdennis** [comment]
It looks like the AVJ format meets most of the requirements I put forth.  It would need to be expanded to include annotations (which don't necessarily have to be populated) and a direct link to the schema source (in the case of `$ref`s).

It wouldn't be too hard for me to implement that.  I just want to nail *something* down before I muck about in my code.  I don't want to start changing it only to have to change it again soon.

**Comment by vearutop** [comment]
https://runkit.com/embed/q7n6kpx6etcv here is a sample error reporting from ajv for ease of consideration.

I see two issues: 
* `schemaPath` only keeps last reference (origins are omitted in case of multiple references)
* the structure is flat, missing hierarchy does not allow to read subschema impact
* (minor) not sure if json pointer should start with `/root`

I think the error response structure is not to replace error messages, I see it as additional interface for programmable processing (not for end-user in any way), so error message phrase/translation/format is out of scope.

Performance-wise implementation could still return true/false for overall validation. Or it could return error response structure right after first root level schema invalidation. To me both of cases are out of scope of response structure format.

Maybe it would make sense to organize error response as a map with data pointers as keys, that would allow to quickly get a list of invalid properties for example for painting it red in some html form.

**Comment by Anthropic** [comment]
@gregsdennis I think dataValue/data should be an optional part of any error object definition, ajv only includes data if verbose output is requested for example and I just use the paths only in my library.

> the structure is flat, missing hierarchy does not allow to read subschema impact

@vearutop ajv's format provides errors for each item even in the subschema (which I feel is too verbose for most needs but great for when debugging), so you can get the full impact can't you, what is missing you feel should be there in your link?

Many frameworks have in Intermediate Representation of the schema (combined with implementation specifics), most likely flatter for machine processing, which is why I would need some convincing on a hierarchy providing benefit, not to say I can't be :trollface:

I mostly agree with the rest :+1:  especially needing both $ref schema paths. Annotations I am not up to speed on their definition, maybe someone can ping me a TL;DR in slack :smile:

**Comment by vearutop** [comment]
@Anthropic please check https://github.com/json-schema-org/json-schema-spec/issues/396#issuecomment-389734109 for hierarchical error response example.

It allows you to understand that on high level that `/root/zoo` failed. It failed because of `#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf` did not match any sub schemas.

If that reason is not enough you can check failure reason further: 0, 1 by enum directly, 2 for a complicated reason.

Then you can continue in the error structure and get that `#/properties/root/patternProperties/^[a-zA-Z0-9_]+$/oneOf/2` failed because it jumped to another schema by `$ref` and then failed in `anyOf` because of all inner subschemas failed by `enum`. 

So, hierarchical error response is kind of story that you can read to get a full understanding of failure reason. Able to quit reading at level that you find deep enough.

With flat list it is not so apparent how to traverse from high level error to its root causes. Looping through whole list you might be able to build hierarchy.

**Comment by gregsdennis** [comment]
I thought it might be beneficial to have an AJV-style output example for the [samples I presented earlier](https://github.com/json-schema-org/json-schema-spec/issues/396#issuecomment-398267008), so here it is.  Note that this isn't AJV's output, it just uses that style.  I've added the additional data that should be there, and consolidated the `message` and `params` properties for readability.

### Example 1 (valid)

*I added 'otherProp' to illustrate a point.*

Instance

```json
{
  "intProp":9,
  "nested":{
  "b":[]
  },
  "otherProp":true
}
```

Output

```json
[
  {
    "dataPath":"/",
    "data":{
      "intProp":9,
      "nested":{
      "b":[]
      }
    },
    "result":"passed",
    "annotations":[
      {
        "source":"/title",
        "value":"root"
      }
    ]
  },
  {
    "dataPath":"/intProp",
    "data":9,
    "result":"passed",
    "annotations":[
      {
        "source":"/properties/intProp/title",
        "value":"it's an int"
      },
      {
        "source":"/properties/intProp/description",
        "value":"found the int"
      }
    ]
  },
  {
    "dataPath":"/nested",
    "data":{
      "b":[]
    },
    "result":"passed",
    "annotations":[
      {
        "source":"/properties/nested/title",
        "value":"internal"
      }
    ]
  },
  {
    "dataPath":"/nested/b",
    "data":[],
    "result":"passed",
    "annotations":[
      {
        "source":"/properties/intProp/title",
        "value":"it's an int"
      },
      {
        "source":"/properties/intProp/description",
        "value":"found the int"
      }
    ]
  }
]
```

Notes

1. Each element that has an annotation needs to have an element in the output array.  This means that the output array can get rather large.
2. I don't mind the idea that the `result` and `data` properties be considered optional so that implementations can in-/exclude this based on a "verbose" flag or similar.  But it's important that the spec defines it so that the output is consistent across implementations.  Also, this format, when containing the data value, would contain multiple copies of portions of the instance, which is where the instance-based organization has a distinct advantage. (@Anthropic)
3. Only those nodes with annotations have entries.  As a result, there's no indication that `otherProp` has no requirements.

### Example 2 (invalid)

Instance

```json
{
  "intProp":9,
  "offsetCoordinate":{
    "X":10,
    "Y":3
  },
  "otherProp":true
}
```

Output

```json
[
  {
    "dataPath":"/",
    "data":3,
    "result":"failed",
    "errors":[
      {
        "source":"/properties/offsetCoordinate",
        "message":"A subschema failed validation."
      }
    ]
  },
  {
    "dataPath":"/offsetCoordinate",
    "data":3,
    "result":"failed",
    "errors":[
      {
        "source":"/properties/offsetCoordinate",
        "message":"A subschema failed validation."
      }
    ]
  },
  {
    "dataPath":"/offsetCoordinate/Y",
    "data":3,
    "result":"failed",
    "errors":[
      {
        "source":"/properties/offsetCoordinate/Y/$ref/minimum",
        "absoluteSource":"/$defs/integerMin5",
        "message":"The value '3' is not greater than or equal to '10'."
      }
    ]
  }
]
```

Notes

1. The first two entries in the output may be optional as their failure is implied.  Technically, they should be present because a failure did occur at those nodes.  Additionally, those elements allow for error messages to be logged for those locations.
2. Only those nodes with failures have entries.  As a result, there's no indication that `otherProp` has no requirements.
3. All annotations are dropped, per annotation requirements.

**Comment by philsturgeon** [comment]
This thread looks like it's really making some progress, great job to everyone for working together on this.

I wonder if we can skip having the `data` field in there, as so far all proposed formats have a JSON Pointer. If folks want to look up the value, they can grab it from there.

**Comment by gregsdennis** [comment]
@philsturgeon one of the things that @handrews is pushing for is to have both the `$ref` path pointer and a direct point that needs no resolution.  This is for convenience of the consumer.

It's following that logic that I keep putting the data value in there. I think having the data present is important.  In that regard, the instance-based output I posted first prevents having duplicated instance data in the output.

**Comment by handrews** [comment]
@philsturgeon @gregsdennis I have not caught up on this thread, and have other things I'm focusing on right now, but I wanted to address the idea of putting the data in the output.

The reason for adding both the direct absolute schema pointer and the a pointer-ish schema path is that both require handling all of the `$id`/`$ref` resolution in the context of performing validation.  It would be so difficult of applications to do this that the feature would be worthless to any application that needed this information.

You need the path to decide which annotation to choose if you have application-specific conventions for such things (e.g. `title` adjacent to the refer overrides `title` in the referred schema).  You need the absolute URI (and it needs to be a URI, not just a pointer) in case you want to look at what else was in that schema for whatever reason.

Applications do not have the same problem with instance data.  Given a JSON Pointer and an instance, it is trivial to resolve the pointer (as in, anyone with access to a JSON Schema library should have a JSON Pointer library available as well, and JSON Pointer isn't exactly hard to implement if you really need to do it yourself).

**Comment by Julian** [comment]
I am as usual unable to stay up with the level of activity here (which is great!), but the one thing I caught in passing here while reading some of these comments I think might be important to point out --

In my implementation, and I suspect many many implementations, I would not implement this schema "directly" as the native error format -- i.e., the result of validating a schema (in success or error) would be more "native" to my language and its norms. So for instance, it is super inconvenient to deal with JSON refs. Way better is to deal with runtime methods that you can call to get the value that the ref points to, and never deal with the ref ever, until you need to say, write it down for later.

So what *would* happen though would be that in my implementation, where something called `ValidationError` is the mechanism here, terminology would start to converge more towards the sort of language used here (which of course is a slow process because I have to maintain backwards compatibility for quite awhile), and perhaps more usefully, I'd add a `ValidationError.as_spec` or some better named thing, which would return the JSON object here.

Just pointing that out because I think it's important for a comment or two here on convenience -- convenience for runtime I'd provide no matter what's in this schema, so I'd be focused more on necessity than runtime convenience.

**Comment by handrews** [comment]
@Julian you're thinking primarily in terms of people using your library's programmatic interface within a single Python process.  That's not really what this issue is about- the specification cannot address the idioms of all programming languages, and of course implementations should be tailored to suit their languages and environments.

But consider a system with an API that is used by the company's web interface using HTML forms with JavaScript, and is also used by customers who call the API directly.

The JavaScript on the web page uses the schema to do client-side validation with a JavaScript implementation.  The API back end could be implemented in Python with your library.  It also needs to do validation because not all requests come from the web site.

Let's also assume that the schema uses at least one custom keyword that does some sort of complicated back-end verification.  It doesn't really matter what it is, it just can't be done in the browser.  That means that the browser might also get validation errors back from the server.

So now the code in the browser has to handle processing and displaying errors from two different implementations in two different programming languages.  One of which is coming over the network.  That means that without an interoperable format, that client code needs to understand multiple ways of presenting the results and errors.  Or the back-end code needs to understand how to translate to whatever the front-end code can use.

Let's further complicate this by assuming a microservice environment, so some API calls are implemented in Go.  Now there are at least three implementations that the front end has to deal with.  And really, the front end and back end should not care which implementations are being used elsewhere.  It should be possible to change languages and implementations in one place without having to update everything else.

To build a real, broadly interoperable ecosystem around JSON Schema, we need to think beyond the idioms and conveniences of individual languages and environments.  They are still important, but they also do not need help from the spec (in fact, the spec trying to mandate idiomatic use would not scale and would generally be a mess).  But the broader ecosystem does need help from the spec, at least at the level of a recommendation.

Implementors should not have to guess about whose output format to mimic, or have to convince each other to support particular formats or concepts out in the wild.  We can have that discussion here and give everyone an interoperability target.

**Comment by Julian** [comment]
That was quite a wide-reaching response for quite a narrow comment :). It of course all makes sense, but I'm not sure what it has to do with my above -- will try and clarify.

I'm in fact very much *not* thinking of just my own implementation or environment, I was coming from exactly the opposite side of "I wouldn't here be aiming for things like 'convenience' as the primary concern when designing this format". I'd aim for standardization and terseness. Convenience can come from something *using* this format to do something.

To be more specific (I'm not sure I even agree with the following, but IIRC it was one of the things that triggered me at least making the above point) -- why should this format include both a pointer *and* the data at that pointer (which from my skim is apparently being proposed). It's true that's something e.g. my library does, but that's for convenience of a runtime user -- why does the 'standard schema for results' want to be anything more than a *minimal* description of the success or failure?

TL;DR: ISTM smaller is better.

Addendum, to try and be useful here too: the kinds of questions that have come up in my own experience here are things like "for the `required` keyword, where is the error? At the missing not-present property, or at the parent?". I go with the latter, and therefore claim to guarantee "all paths you find in errors are resolvable against the instance", but I suspect others might have made other choices. Same goes for "where is the error for additional unexpected properties that were disallowed by `additionalProperties`?"

And on another route, "what info is provided for errors from the `format` validator, and where does specific information about them go"

**Comment by handrews** [comment]
@Julian my apologies, I misread your comment as something of a "putting stuff in a standard form like this isn't that useful because I'm just going to make a different interface for Python anyway."  I see that that was incorrect.  I will follow up on your other points later.

**Comment by Julian** [comment]
Ah yes! Very much not that -- definitely more:

If you make a small standard thing, we can all provide that as an interchange, but need not worry about making that small standard thing all-encompasing, because convenience is well, in the eye of the beholder and very context-driven.

(And no worries! Definitely didn't assume it was hostile :)

**Comment by gregsdennis** [comment]
It sounds like we don't want to include the instance data in the results. In that case, the flat list format with instance pointers for

- the instance location
- the schema `$ref` path
- the resolved schema

will be fine for the spec.

Furthermore, when performing a simple validation, and the instance passes, it should be sufficient to respond with `true`. However, when annotations are requested, I suggest we follow a similar format as with error output of for no other reason than consistency.

**Comment by handrews** [comment]
I've updated PR #610, which covers the annotation collection process and requirements for data collected, but without specifying the exact format.  It now also has an example of how some of that information is intended to be used by applications.

@gregsdennis your most recent comment sounds great, except that the resolved schema needs to be a URI, not just a pointer.  Schema locations may come from multiple schema documents, so a pointer alone is inadequate.

The schema path can be a pointer, although we'll have to be clear about it being a pointer for the post-resolution data structure (or some such terminology).

**Comment by Anthropic** [comment]
I find the goal of the output is to provide a guide to the most _efficient, implementable, minimal and extensible output_ for a foundation, that's why hierarchy didn't sit so well with me, recursion being more expensive than flat loops.

@gregsdennis given your suggestion of a simple boolean result do you feel that perhaps the result should be within an object to provide the overall output?

It goes against the mantra I just wrote of being minimal, but it could also make it easier to consistently expect the `valid` property to exist, but also to be able to know when there are errors vs annotations in just an array result. The annotation and validation processes still remain quite separate.

Do you see the end array result combined or under separate keywords?

```js
// Minimum
{
  "valid": true
}

// Annotation focused:
{
  "valid": true,
  "annotation": [...]
}

// Errors and/or Annotations:
{
  "valid": false,
  "errors": [...],
  "annotations": [...]
}
// or
{
  "valid": false,
  "results": [...errors, ...annotations]
}
```

**Comment by gregsdennis** [comment]
That seems logical. 🖖

**Comment by vearutop** [comment]
I'm still in favor of hierarchical structure with variable/optional level of deepness (say verbosity).

And also what do you think about using data pointers strings as keys in map of error instead of having arrays of errors? That would allow to check if particular field in data has failed or passed without iterating.

**Comment by gregsdennis** [comment]
@vearutop I explored using pointers as keys. The problem is that a single field can fail with multiple errors. You'd have to array the errors in the end, anyway.

While I do like the instance-based approach, I'm finding it difficult to figure out how I'd implement it.  Also, the majority favors the flat format, and a few existing implementations (mine included) already use it to some extent.

**Comment by yurikhan** [comment]
I think we should not choose an approach by ease of implementation only. We should also strive to make the validation result actually useful for the client.

I’m not convinced how a flat format would allow the consumer to track down the exact chain of reasons why validation failed. The example validation result linked in https://github.com/json-schema-org/json-schema-spec/issues/396#issuecomment-400553752 is really unhelpful in that.

**Comment by gregsdennis** [comment]
> @yurikhan a flat format would allow the consumer to track down the exact chain of reasons why validation failed

I see where you're coming from, and I think that omitting errors like `A subschema failed validation.` (from my [invalid flat example](https://github.com/json-schema-org/json-schema-spec/issues/396#issuecomment-401177510)) may help clear some of that up.  In the flat format those errors admittedly hide the real issue, which is (in this case) `The value '3' is not greater than or equal to '10'.`

**Comment by yurikhan** [comment]
That is so only in the simplest cases, namely, logical conjunction of conditions.

Consider a moderately complex schema involving a couple of disjunctions:

```json
{"type": "object", "properties": {
  "foo": {"type": "number", "anyOf": [
    {"maximum": 5},
    {"minimum": 15}
  ]},
  "bar": {"type": "integer", "anyOf": [
    {"multipleOf": 3},
    {"multipleOf": 5}
  ]}
}}
```

The value `{"foo": 10, "bar": 2}` is invalid against this schema, for 4 atomic reasons.

A hierarchical response will say (paraphrasing from json): “it is invalid _because_ `#/foo` (10) is _neither_ less than 5 _nor_ greater than 15, _and_ `#/bar` (2) is _neither_ a multiple of 3 _nor_ a multiple of 5”. See, in this simple case, I didn’t even have to quote schema references.

A flat response will say… what?

* “According to `#/properties/foo/anyOf/0/maximum`, `#/foo` (10) should be less than 5;
* according to `#/properties/foo/anyOf/1/minimum`, `#/foo` (10) should be greater than 15;
* according to `#/properties/foo/anyOf`, `#/foo` (10) should satisfy at least one subschema;
* according to `#/properties/bar/anyOf/0/multipleOf`, `#/bar` (2) should be a multiple of 3;
* according to `#/properties/bar/anyOf/1/multipleOf`, `#/bar` (2) should be a multiple of 5;
* according to `#/properties/bar/anyOf`, `#/bar` (2) should satisfy at least one subschema”?

Imagine just a few more properties, and, as a human reading that list, I have to resort to pencil and paper to reconstruct the hierarchy. Understanding the exact nature of the violation(s) from the validation response is marginally easier than just manually re-validating against the schema; and I honestly don’t see how to write code that turns that list into a message suitable for presenting to an end user (who knows their way around JSON but not necessarily a Schema wizard).

And we have not even broken out `if`/`then`/`else` yet, which are mostly redundant syntax sugar in terms of “valid/invalid” boolean logical expressive power, but are theoretically capable of getting us a long way towards more understandable “invalid because” violation reports.

**Comment by gregsdennis** [comment]
The flat response is only flat in regard to listing the JSON instance values.  Within each, the annotations/errors are grouped;  we don't have an entry for each data/result combination.

So the flat response will say...

- `/foo` is invalid because
    - the value should meet any of
        - the value should be less than 5
        - the value should be more than 10
- `/bar` is invalid because
    - the value should meet any of
        - the value should be a multiple of 3
        - the value should be a multiple of 5

So, yeah, there has to be some hierarchy to account for `*Of`.  But we group by the data location.

---

I like this.  We're getting closer...

**Comment by gregsdennis** [comment]
You're still grouping by data location.

- `/foo` is invalid because
    - the value must be in the set `["a"]` (from schema location `/anyOf/0/properties/foo/enum`)
- `/bar` is invalid because
    - the value must be a multiple of 2 (from schema location `/anyOf/1/properties/bar/multipleOf`)

*If* you wanted to include that the entire thing fails because `/anyOf` didn't match any of the subschemas, then I would expect the hierarchical layout to be better.  But if you can infer that that subschema failed from the errors above, then I think the list would be better.

**Comment by vearutop** [comment]
I think grouping by data location makes sense because naturally JSON Schema does not allow cross data references (at least so far) and every violation is tightly coupled with some exact place in data.

```
{
 "/somewhere/in/data": [<violations>],
 "/another/place/in/data": [<violations>]
}
```

**Comment by gregsdennis** [comment]
I'm still not very comfortable with pointers as keys, but it *is* more terse.  It probably just comes from my experience in .Net.  I like keys to have valid property names.

And I do like the idea of having an encompassing object that holds the overall result and a list of annotations/errors.

**Comment by vearutop** [comment]
@gregsdennis yes, now I agree, https://github.com/json-schema-org/json-schema-spec/issues/396#issuecomment-402051319 this example shows very well that you can not always split violations in a map

**Comment by yurikhan** [comment]
Every violation is coupled with an exact place, yes. But each violated subschema has a different role in the validity of the root instance.

Let me construct an example where a single location participates via two different paths:

```json
{"anyOf": [
  {"properties": {
    "foo": {"enum": ["a"]},
    "bar": {"multipleOf": 3}
  }},
  {"properties": {
    "foo": {"enum": ["b"]},
    "bar": {"multipleOf": 5}
  }}
]}
```

No matter how you slice it, grouping that by location is not going to be helpful.

**Comment by gregsdennis** [comment]
@yurikhan that's a good example.

Instance data of `{"foo":"a","bar":10}` would fail, but how do you indicate the failure?

- `foo` passes on `/anyof/0`, but then `bar` fails there
- `bar` passes on `/anyof/1`, but then `foo` fails there

The full instance only passes on one portion of each of the subschemas.  To do this as a list would be

- `/`
    - at least one subschema should pass (from `/anyOf`)
- `/foo`
    - value must be in the set '["b"]' (from `/anyOf/1/foo`)
- `/bar`
    - value must be a multiple of 3 (from `/anyOf/0/bar`)

Without the first item, there's no way to indicate that only one of these requirements needs to be met.

Maybe that's how the hierarchy helps.  That would result in

- `/`
    - at least one subschema should pass (from `/anyOf`)
    - data
        - `/foo`
            - value must be in the set '["b"]' (from `/anyOf/1/foo`)
        - `/bar`
            - value must be a multiple of 3 (from `/anyOf/0/bar`)

This at least includes the result from the `anyOf` in a way that indicates some sort of relationship.

---

This brings us back to the discussion of "do we base the results on the schema or the instance?"  I think this example shows that grouping based on either leaves the other lacking organization.  If that's the case, maybe we shouldn't group at all: just have a list of objects where each object represents a single annotation/error of a single data point.  Then the client can decide how to group them.

```
[
  {
    "data":"/",
    "keyword":"/anyOf",
    "message":"at least one subschema should pass"
  },
  {
    "data":"/bar",
    "keyword":"/anyOf/0/bar",
    "message":"value must be a multiple of 3"
  },
  {
    "data":"/foo",
    "keyword":"/anyOf/1/foo",
    "message":"value must be in the set ["b"]"
  },
]
```

This should make everyone equally unhappy.

**Comment by yurikhan** [comment]
> Instance data of `{"foo":"a","bar":10}` would fail, but how do you indicate the failure?

For completeness, could you please analyze `{"foo": "c", "bar": 8}` case? I think it’s even more instructive than the one above which satisfies half of each path.

> This should make everyone equally unhappy.

I will refrain from commenting on this metric of success.

**Comment by yurikhan** [comment]
See, this is why I’m opposed to a format that tries to structure itself after the instance: It cannot preserve the essential semantics of many validation results.

On the other hand, in a hierarchical format structured after the schema, it’s straightforward. For `{"foo": "c", "bar": 8}`:

* `/` does not satisfy `/` because it should satisfy at least one of two subschemas of `/anyOf`, but:
  * `/` does not satisfy `/anyOf/0` because:
    * `"c"` (`/foo`) is not any of `["a"]` (`/anyOf/0/properties/foo/enum`), and
    * `8` (`/bar`) is not a multiple of `3` (`/anyOf/0/properties/bar/multipleOf`)
  * `/` does not satisfy `/anyOf/1` because:
    * `"c"` (`/foo`) is not any of `["b"]` (`/anyOf/1/properties/foo/enum`), and
    * `8` (`/bar`) is not a multiple of `5` (`/anyOf/1/properties/bar/multipleOf`)

For `{"foo": "a", "bar": 10}:`

* `/` does not satisfy `/` because it should satisfy at least one of two subschemas of `/anyOf`, but:
  * `/` does not satisfy `/anyOf/0` because:
    * `10` (`/bar`) is not a multiple of `3` (`/anyOf/0/properties/bar/multipleOf`)
  * `/` does not satisfy `/anyOf/1` because:
    * `"a"` (`/foo`) is not any of `["b"]` (`/anyOf/1/properties/foo/enum`)

**Comment by gregsdennis** [comment]
This brings up the reason I originally proposed the instance-oriented format.  I searched for it, and I fear it's likely in a Slack discussion somewhere, but the idea was that a schema-oriented format would require the client to scan for errors that occurred due to a single instance location.

Like I said before, the application is going to have to do some level of analysis.  If we choose an instance-oriented output, the application will have to scan for schema locations.  If we choose a schema-oriented output, the application will have to scan for instance locations.

It's a matter of perspective and what the client deems important, I suppose.  I don't know that we can make that decision for all clients.

---

There is an alternative: we could require (or hard-suggest) that implementation support both formats.  I'm not terribly excited about that, though.

**Comment by yurikhan** [comment]
> a schema-oriented format would require the client to scan for errors that occurred due to a single instance location.

What’s the use case there? Do you want to be able to tell the user, “the value of this property here is invalid”? I’m afraid you cannot reliably do that.

With the full expressive power of JSON Schema, the value of a single location in the instance is not invalid by itself. Rather, a specific combination of multiple locations of the instance makes the whole instance invalid, and you cannot point your finger and say which of them did that, because changing any of them could make the instance valid again.

(Well, if you avoid using `anyOf`, `oneOf`, `not`, `dependencies`, or `if`/`then`/`else`, then you might be able to talk about each specific location being valid or invalid. But that limits the usefulness of schema validation severely.)

**Comment by erosb** [comment]
TL;DR & leaving some notes here:

 * the error reporting format of my implementation is [documented here](https://github.com/everit-org/json-schema#investigating-failures) - just FYI. It is close to the hierarchy format proposed above. It includes both instance pointer and schema pointer.
 * I'm torn about including instance data in the standard validation error format (at least make it optional, please), mainly because we don't have control over the size of the invalid data, and it may result in possible performance problems for users.

**Comment by gregsdennis** [comment]
@yurikhan I think in the end I was just looking to suggest an alternative approach and provoke discussion.

If we decide to go with the schema-based approach, we will definitely need to figure out how to represent `$ref` locations within the output structure.  The instance-based approach doesn't have that need; instead they are handled via the `source` pointer and `directSource` URI in the result object.

**Comment by gregsdennis** [comment]
Thank you @Anthropic.  I completely agree that we should focus on applications as the primary consumers.  Applications can then transform the results to whatever format is best suitable for their users.  If implementations want to offer alternate formats (grouping, etc.), that is their prerogative.  But as a minimum, the format should be easily consumable by a computer.  I agree that an ungrouped list facilitates that best.

**Comment by epoberezkin** [comment]
I agree with @Anthropic, I can only add that:
1) development time/cost is in most cases a higher priority than performance - and it will only be more so going forward.
2) for the majority of use cases array is sufficient, and given that it’s simpler to create and to iterate with any language primitives than trees it is a better choice, even if constructing tree from array can be marginally more complex.

The list of errors from validation can be seen as something similar to an exception stack trace - a linear representation of the tree of functions that led to an error, from the inner to the outer. I don’t remember anybody complaining about the problem with linear stack traces.

**Comment by yurikhan** [comment]
@epoberezkin 
You don’t have problems with linear stack traces *because* they are linear, and stack traces are linear because they reflect the state of a single thread of execution. If you have a function `foo` trying to call `bar`, falling back to `baz` and then to `quux`, and succeeding if any of those succeed, then you only get the exception and stack trace from `quux`.

An `anyOf` schema keyword, on the other hand, can be thought of as forking to execute all of subschemas in parallel, then joining and reporting all exceptions if no branch succeeded.

**Comment by gregsdennis** [comment]
> @yurikhan In my experience, flattening a hierarchy is almost always easier to implement than reconstructing it from a flat list.

I understand that you're speaking from your experience, but you can't generalize this kind of thing.

Consider that if we end up using a schema-based hierarchical format, but the client wants an instance-based hierarchical format, they first have to linearize the result before they can restructure it.  The same is true for the reverse.

But if we use a list format, those who want either hierarchy potentially have one fewer step to work out.  And it shouldn't be a difficult task for implementations to output a list considering that many already do.

**Comment by Anthropic** [comment]
> In my experience, flattening a hierarchy is almost always easier to implement than reconstructing it from a flat list. Could you please walk us through an example of such reconstruction? (Especially in the face of $ref in subschemas.)

@yurikhan I suspect there is a communication issue here, my [previous post](https://github.com/json-schema-org/json-schema-spec/issues/396#issuecomment-402575647) already contains enough information to conclude why that is not relevant.

I'd like to ensure you are comfortable with the outcome, can you please message me in Slack so I can work through it with you?

**Comment by vearutop** [comment]
@Anthropic I must disagree, this thread has plenty of reason why hierarchy IS relevant.

I don't think postprocessing is trivial, at least you will have to have JSON Pointer parser utility at hands. 

I also don't think performance aspect is important here, as I consider having this response optional upgrade from true/false validation status (it is not a must to create such error response for every validation session).

Also, having the spec for format does not prevent implementation from having optional optimizations. Like for example "do not collect logical subschema errors to the response", in that case you would end up with nice flat array. If the consumer is not interested in deeper details why would you collect subschema errors anyway? And on the opposite why would the consumer build hierarchy himself rather than just get and use it in spec format?

**Comment by yurikhan** [comment]
@Anthropic (Re: Slack) No, I’d much prefer to keep discussion public.

_As a_ producer of instances that may be invalid and an end user of a JSON Schema validator, _I want_ a readable error message, _so that_ by reading it I immediately know what is wrong with the instance (and thus my code that generated it). In cases when that cannot be decided automatically, I want the error message to contain enough structure that I could quickly dismiss irrelevant violations and focus on the relevant part. (In the “(`foo` equal to `"a"` and `bar` multipleOf `3`) or (`foo` equal to `"b"` and `bar` multipleOf `5`)”  example with instance `{"foo": "a", "bar": 10}` presented upthread, I probably know I meant the `"a"` case, but the implementation cannot know that.)

_As an_ application developer, _I want_ to be able to easily generate error messages of the kind described above, _so that_ possible bugs in the validation report to error message transformation code cannot cloud the real issue. Ideally, I would achieve that by doing a node-to-node, structure-preserving transformation from a hierarchical validation report object to a hierarchical error message object; but I am willing to settle for having to reconstruct the hierarchy from a flat list if it is demonstrated to be sufficiently easy.

As a validation library contributor, I know the desired structure is not very hard or expensive to maintain within the implementation. I further know that structure naturally emerges as a consequence of the validation logic.

As a potential specification contributor, I note that, until now, all validation has been defined recursively (“An instance validates successfully against this keyword [`anyOf`] if it validates successfully against at least one schema defined by this keyword’s value”). It will be straightforward to extend that if the validation result format is hierarchical with structure based on the schema, leading to a specification that is not much harder to comprehend than its current version.

**Comment by Relequestual** [comment]
@yurikhan @Anthropic It's perfectly fine to have off-issue discussions, and then come back to document the result of the discussion. I personally find this preferable to mega-theads, which then become unwhealdy to anyone who hasn't kept up *raises hand*. Although, I also respect if you don't want to use slack at all. longer discussions with summary comments with the other party agreeing on the summary is nicer when there are issues of confusion or dissagreement =]

**Comment by Relequestual** [comment]
My 2c, the errors presented to the user by https://www.jsonschemavalidator.net are pretty handy for schema debugging, including the attachment to the instance json!
I haven't looked into how it's represented in code, but that implementation should be considered as an example which is functionally useful.

**Comment by Relequestual** [comment]
If discussion is going to continue here, I would expect that once a general consensus can be reached, a new summary can be created (either at the end or as a new issue). I'll defer to @handrews on this.
A reminder: general consensus doesn't mean that everyone must agree, but generally it looks like everyone was understood, and most people seem to agree on a direction to move on so a PR can be started. PR will get feedback to make it tighter, and that's OK; we don't need to nail it down completely here and now.

**Comment by vearutop** [comment]
For quick reference, here is sample result output of https://www.jsonschemavalidator.net/
`{"foo":"f","bar":12}` validated with https://github.com/json-schema-org/json-schema-spec/issues/396#issuecomment-402069476
```json
{
  "valid": false,
  "schemaErrors": [],
  "validationErrors": [
    {
      "message": "JSON does not match any schemas from 'anyOf'.",
      "lineNumber": 1,
      "linePosition": 20,
      "path": "",
      "value": null,
      "schemaId": "#",
      "schemaBaseUri": null,
      "errorType": "anyOf",
      "childErrors": [
        {
          "message": "Value \"f\" is not defined in enum.",
          "lineNumber": 1,
          "linePosition": 10,
          "path": "foo",
          "value": "f",
          "schemaId": "#/anyOf/1/properties/foo",
          "schemaBaseUri": null,
          "errorType": "enum",
          "childErrors": []
        },
        {
          "message": "Value \"f\" is not defined in enum.",
          "lineNumber": 1,
          "linePosition": 10,
          "path": "foo",
          "value": "f",
          "schemaId": "#/anyOf/0/properties/foo",
          "schemaBaseUri": null,
          "errorType": "enum",
          "childErrors": []
        },
        {
          "message": "Integer 12 is not a multiple of 5.",
          "lineNumber": 1,
          "linePosition": 19,
          "path": "bar",
          "value": 12,
          "schemaId": "#/anyOf/1/properties/bar",
          "schemaBaseUri": null,
          "errorType": "multipleOf",
          "childErrors": []
        }
      ]
    }
  ],
  "schemaParseError": null,
  "jsonParseError": null
}
```

**Comment by vearutop** [comment]
It would be nice to come up with a rich example of schema and data, having multiple dereferences, `not`, `oneOf` to also see the case when missing failure is invalid. And then to assert how helpful are currently implemented error systems.

**Comment by Relequestual** [comment]
@vearutop thanks for the example, and your last comment.
I was going to suggest something similar, a more analytical approach of what people currently do, and if there's an superset of those implementations which cover as many requirements as are defined to satisfy this issue... probably need to define the requirements clearly also. BUT I realised that was a lot of work for someone to do, and I know I don't have the time.

**Comment by vearutop** [comment]
I've tried to make a sample schema and response output, please check: 
https://gist.github.com/vearutop/376e2a148e6df044417264722f0d1319

Same schema and test cases with `ajv` processing:
https://runkit.com/embed/w5fpvewlnetl

The idea is to provide a [flat response](https://gist.github.com/vearutop/376e2a148e6df044417264722f0d1319#file-test_min-json) (array of causes) for top level schema. 
Optionally subschema can produce [a recursive array of causes](https://gist.github.com/vearutop/376e2a148e6df044417264722f0d1319#file-test-json) in `causeContext`.

A single cause is a structure of
* keyword (not present for boolean schema)
* schemaPath
* dataPath
* valid (true/false)
* causeContext ( a list of causes where schema of current cause acts as a top level schema)

For multiple schema keywords (`*Of`), each element creates a separate `causeContext`.

References [create](https://gist.github.com/vearutop/376e2a148e6df044417264722f0d1319#file-test-json-L49) a new `causeContext` allowing you to track `$ref` jumps.

I don't like having all causes in a single flat list because non top level causes add noise but does not help you to analyze the root cause until you rebuild flat list into hierarchy.

**Comment by expelledboy** [comment]
Hi guys, I am new to json schema, implementing a validation library in Erlang. Very interested in this discussion as a framework developer.

As a suggestion to reach consensus I suggest we summarize the points made here into an Enhancement Proposal, eg in a project wiki page? I have read through this mega-thread and conclusions have already been made, which detailed concisely might align the community.

**Comment by gregsdennis** [comment]
@expelledboy that's a good idea, and it's been suggested before.  I'll try to write up a new issue that summarizes where we are with this one.  Then we can close this out.

**Comment by Anthropic** [comment]
@vearutop I agree with the idea of the `default` output and the `causality` output providing the same detail at the root error array so there is no difference in error counts or overall behaviour between the two.

@gregsdennis @vearutop 
I feel like we could consider:
1. **basic (default)**: the flat array minimal output
2. **causality**: as basic, but with @vearutop 's `causeContext`
3. **verbose**: a hierarchy

Then for each level the **Annotation/Error Description Object** could be optionally alternated with an  **_Enhanced_ Annotation/Error Description Object** with more detail, message ids, data, embedded schemas, etc... to be decided... this gives implementations a lot of choice while not requiring anyone to go too far from their comfort zone until enough people request it of them.

I feel an implementation:
SHOULD provide basic
RECOMMENDED to provide causality and
OPTIONALLY provide verbose
**only in so far as being able to claim they "support the standard annotation/error output"**. I like the idea of having a minimum requirement before someone can make that claim.

**Comment by expelledboy** [comment]
@gregsdennis Could I suggest rather just starting a wiki page, and continuing the discussion here? We may just end up with another mega-thread.

**Comment by gregsdennis** [comment]
@expelledboy I'm not sure that's in line with the organization of the repo.  Also, this topic is very much in discussion.  I'll discuss it over on Slack with the powers that be here.

**Comment by retrosight** [comment]
FWIW, for the purposes of troubleshooting what went wrong and providing the right level of information to the developer (balancing simplicity with verbosity) we settled on this:

**Comment by vearutop** [comment]
Let's also discuss how we can expose that `required` property is missing.

My proposal based on previous structure is:
For schema `{"required": ["foo", "bar"]}` and data `{"baz":1}` to have this response
```json
{
  "valid": false,
  "causeContext": [
    {"keyword": "required", "schemaPath": "#/required", "dataPath": "#", "valid": false, "causeContext": [
      {"dataPath": "#/foo", "valid": false},
      {"dataPath": "#/bar", "valid": false}
    ]}
  ]
}
```
or this response
```json
{
  "valid": false,
  "causeContext": [
    {"keyword":"required", "schemaPath": "#/required", "dataPath": "#/foo", "valid": false},
    {"keyword":"required", "schemaPath": "#/required", "dataPath": "#/bar", "valid": false}
  ]
}
```

Not sure which one would be better.

**Comment by gregsdennis** [comment]
I don't know why you'd treat `required` differently than any other keyword.  It's just a messaging question.  "One or more required properties were not found" should be sufficient.  If the implementation wished to list the properties, that would be fine, too.

This is about the structure, though, not standardized messaging.

**Comment by vearutop** [comment]
~~But to me list of missing required properties is essential information, that's why I think `required` may need a special treatment.~~

~~If `required` was an object (like `properties`), there would be no problem to expose, as every missing property would have an own `schemaLocation`. But since `required` is an array.~~

I think I just got a rubber duck epiphany, `schemaLocation` can have failed property index :)
```
{
  "valid": false,
  "causeContext": [
    {"keyword":"required", "schemaPath": "#/required/0", "dataPath": "#/foo", "valid": false},
    {"keyword":"required", "schemaPath": "#/required/1", "dataPath": "#/bar", "valid": false}
  ]
}
```

**Comment by gregsdennis** [comment]
Closing this issue in favor of the summary. ☝️ (Please read it.)

**Comment by Relequestual** [comment]
Closed in favour of #643 as requested

**Comment by gregsdennis** [comment]
@expelledboy the preference is that the discussion continues on the new issue.

**Comment by KayEss** [comment]
I had to put together something for this when we were doing our error reporting for our APIs so did some thinking about this and came up with something very like the flat structure above. A few things that I was thinking about though.

First, our validator returns a data structure describing the annotations or the first error (we stop on first at the moment). It does not produce something that is JSON, so maybe the wording needs to be something like "If the validator reports errors as JSON then it SHOULD use a structure like..."

Second, the only things that are really needed are the data URL and the schema URL. The error report doesn't need to include instance data (the client knows it already) or the schema itself (either the client already knows it, or if it doesn't then it's not the APIs job to disclose it).

Thirdly, we also put in a helper tag for the error. This is useful for testing as you can see that the assertion that failed was the one you expected to fail. We only use the schema keyword though, and this is to better support localisation. The problem with something like:

    "message": "JSON does not match any schemas from 'anyOf'."

Is do you translate it? If so, does that mean the validator that produces needs to understand your localisation mechanism? (ouch) If localisation is done on the client then that's a rather long string to be looking up with and very fragile as there'll be a temptation to keep changing it to produce better error messages. I think this whole thing should be up to the consumer/client and all it really wants is a hint as to what failed. For that purpose then just having something like `"failed": "anyOf"` is actually the best thing to use.
